File: AVM/Memory.cs
using System.Text;

namespace AVM;

public class Memory
{
    public int basePointer;
    public int stackPointer;
    public int heapPointer;
    
    public readonly int stackSize = 512;
    
    private byte[] bytes;
    private MemoryLogger logger;

    public string Log => logger.ToString();

    public Memory()
    {
        bytes = new byte[1024];

        basePointer = stackPointer = 0;
        heapPointer = stackSize;

        logger = new(this);
    }

    public int Allocate_Stack(int bytesToAllocate)
    {
        int pointer = stackPointer;

        logger.Log_Allocate(bytesToAllocate);
        
        stackPointer += bytesToAllocate;
        if (stackPointer >= stackSize)
        {
            throw new Exception($"StackOverFlow stack pointer = {stackPointer}, stackSize = {stackSize}");
        }
        
        return pointer;
    }

    public int Allocate_Heap(int bytesToAllocate)
    {
        logger.Log_AllocateHeap(bytesToAllocate);
        
        int pointer = heapPointer;
        heapPointer += bytesToAllocate;
        return pointer;
    }

    public void Deallocate_Stack(int bytesToDeallocate)
    {
        logger.Log_Deallocate(bytesToDeallocate);
        stackPointer -= bytesToDeallocate;
    }

    public void Write(int address, byte value)
    {
        if (address < 0 || address >= bytes.Length)
        {
            throw new Exception($"Write at {address} out of memory bounds ({bytes.Length})");
        }

        logger.Log_Write(address, value);
        
        bytes[address] = value;
    }

    public void WriteShort(int address, short value)
    {
        Write(address, BitConverter.GetBytes(value));
    }
    public void WriteInt(int address, int value)
    {
        Write(address, BitConverter.GetBytes(value));
    }
    public void WriteLong(int address, long value)
    {
        Write(address, BitConverter.GetBytes(value));
    }
    public void Write(int address, byte[] value, bool noLogs = false)
    {
        if (address < 0 || address + value.Length >= bytes.Length)
        {
            throw new Exception($"Write at {address}..{address + value.Length} out of memory bounds ({bytes.Length})");
        }
        
        if (!noLogs) logger.Log_Write(address, value);
        
        for (int i = 0; i < value.Length; i++)
        {
            bytes[address + i] = value[i];
        }
    }

    public byte Read(int address)
    {
        if (address < 0 || address >= bytes.Length)
        {
            throw new Exception($"Read at {address} out of memory bounds ({bytes.Length})");
        }
        return bytes[address];
    }
    public short ReadShort(int address)
    {
        return BitConverter.ToInt16(bytes, address);
    }
    public int ReadInt(int address)
    {
        return BitConverter.ToInt32(bytes, address);
    }
    public long ReadLong(int address)
    {
        return BitConverter.ToInt64(bytes, address);
    }

    public byte[] Read(int address, byte sizeInBytes)
    {
        if (address < 0 || address + sizeInBytes >= bytes.Length)
        {
            throw new Exception($"Read at {address}..{address + sizeInBytes} out of memory bounds ({bytes.Length})");
        }
        
        byte[] value = new byte[sizeInBytes];
        for (int i = 0; i < sizeInBytes; i++)
        {
            value[i] = bytes[address + i];
        }
        return value;
    }

    public void PushInt(int value)
    {
        Push(BitConverter.GetBytes(value));
    }
    public void Push(byte[] bytes)
    {
        logger.Log_Push(bytes);
        
        Write(stackPointer, bytes, true);
        stackPointer += bytes.Length;
    }

    public int PopInt()
    {
        return BitConverter.ToInt32(Pop(sizeof(int)));
    }
    public byte[] Pop(byte bytesToPop)
    {
        logger.Log_Pop(bytesToPop);
        
        stackPointer -= bytesToPop;
        return Read(stackPointer, bytesToPop);
    }


    public int ToAbs(int rbpOffset)
    {
        return basePointer - rbpOffset;
    }

    public void Dump(string filepath)
    {
        File.WriteAllBytes(filepath, bytes);
    }
}

File: RustVM/src/vm/memory.rs
pub struct Memory
{
    pub bytes: [u8; Memory::STACK_SIZE as usize],
    pub stack_pointer: i32,
    pub base_pointer: i32,
    pub heap_pointer: i32,
    pub data_section_size: i32,
}

impl Memory
{
    pub const STACK_SIZE: i32 = 512;

    pub fn new() -> Self
    {
        Self {
            bytes: [0; Memory::STACK_SIZE as usize],
            stack_pointer: 0,
            base_pointer: 0,
            heap_pointer: Memory::STACK_SIZE / 2,
            data_section_size: 0,
        }
    }

    pub fn allocate_stack(&mut self, bytes_to_allocate: i32) -> i32
    {
        let pointer = self.stack_pointer;

        self.stack_pointer += bytes_to_allocate;
        if self.stack_pointer >= Memory::STACK_SIZE
        {
            panic!("Failed to allocate {bytes_to_allocate} bytes on stack due to stack overflow")
        }

        pointer
    }
    pub fn allocate_heap(&mut self, bytes_to_allocate: i32) -> i32
    {
        let pointer = self.heap_pointer;
        self.heap_pointer += bytes_to_allocate;
        pointer
    }
    pub fn deallocate_stack(&mut self, bytes_to_deallocate: i32)
    {
        self.stack_pointer -= bytes_to_deallocate;
    }

    pub fn push_int(&mut self, value: i32)
    {
        self.write_int(self.stack_pointer, value);
        self.stack_pointer += 4;
    }
    pub fn pop_int(&mut self) -> i32
    {
        self.stack_pointer -= 4;
        self.read_int(self.stack_pointer)
    }

    pub fn write_slice(&mut self, address: i32, bytes: &[u8])
    {
        self.slice(address, bytes.len() as i32).copy_from_slice(bytes);
    }
    pub fn write_vec(&mut self, address: i32, bytes: Vec<u8>)
    {
        self.slice(address, bytes.len() as i32).copy_from_slice(&*bytes);
    }
    pub fn write_int(&mut self, address: i32, value: i32)
    {
        self.slice(address, 4).copy_from_slice(value.to_ne_bytes().as_slice());
    }
    pub fn write_byte(&mut self, address: i32, value: u8)
    {
        self.bytes[address as usize] = value;
    }

    pub fn read(&self, address: i32, count: i32) -> &[u8]
    {
        &self.bytes[(address as usize)..((address + count) as usize)]
    }
    pub fn slice(&mut self, address: i32, count: i32) -> &mut [u8]
    {
        &mut self.bytes[(address as usize)..((address + count) as usize)]
    }
    pub fn read_int(&self, address: i32) -> i32
    {
        i32::from_ne_bytes(self.read(address, 4).try_into().unwrap())
    }

    pub fn copy(&mut self, src_address: i32, dst_address: i32, count: i32)
    {
        let src_slice = self.read(src_address, count).to_vec();
        self.slice(dst_address, count).copy_from_slice(&*src_slice);
    }

    pub fn to_abs(&self, rbp_offset: i32) -> i32
    {
        self.base_pointer + rbp_offset
    }
}